attributing-content.md

```diff
@@ -0,0 +1,121 @@
+the unattributed content as-is, but it will render the attributed content with
+the additional information. As all changes in Yjs are identifyable by Ids, we
+can use `IdMap`s to map changes to "attributions". For example, we could
+attribute deletions and insertions of a change and render them:
+
+```js
+// We create some initial content "Hello World!". Then we create another
+// document that will have a bunch of changes (make "Hell" italic, replace "World"
+// with "Attribution").
+const ydocVersion0 = new Y.Doc({ gc: false })
+ydocVersion0.getText().insert(0, 'Hello World!')
+const ydoc = new Y.Doc({ gc: false })
+Y.applyUpdate(ydoc, Y.encodeStateAsUpdate(ydocVersion0))
+const ytext = ydoc.getText()
+ytext.applyDelta([{ retain: 4, attributes: { italic: true } }, { retain: 2 }, { delete: 5 }, { insert: 'attributions' }])
+// this represents to all insertions of ydoc
+const insertionSet = Y.createInsertionSetFromStructStore(ydoc.store)
+const deleteSet = Y.createDeleteSetFromStructStore(ydoc.store)
+// exclude the changes from `ydocVersion0`
+const insertionSetDiff = Y.diffIdSet(insertionSet, Y.createInsertionSetFromStructStore(ydocVersion0.store))
+const deleteSetDiff = Y.diffIdSet(deleteSet, Y.createDeleteSetFromStructStore(ydocVersion0.store))
+// assign attributes to the diff
+const attributedInsertions = createIdMapFromIdSet(insertionSetDiff, [new Y.Attribution('insert', 'Bob')])
+const attributedDeletions = createIdMapFromIdSet(deleteSetDiff, [new Y.Attribution('delete', 'Bob')])
+// now we can define an attribution manager that maps these changes to output. One of the
+// implementations is the TwosetAttributionManager
+const attributionManager = new TwosetAttributionManager(attributedInsertions, attributedDeletions)
+// we render the attributed content with the attributionManager
+let attributedContent = ytext.getContent(attributionManager)
+console.log(JSON.stringify(attributedContent.toJSON().ops, null, 2))
+let expectedContent = delta.create().insert('Hell', { italic: true }, { attributes: { italic: ['Bob'] } }).insert('o ').insert('World', {}, { delete: ['Bob'] }).insert('attributions', {}, { insert: ['Bob'] }).insert('!')
+t.assert(attributedContent.equals(expectedContent))
+
+// this is how the output would look like
+const output = [
+  {
+    "insert": "Hell",
+    "attributes": {
+      "italic": true
+    },
+    "attribution": {                // no "insert" attribution: the insertion "Hell" is not attributed to anyone
+      "attributes": {
+        "italic": [                 // the attribute "italic" was added by Bob
+          "Bob"
+        ]
+      }
+    }
+  },
+  {
+    "insert": "o "                  // the insertion "o " has no attributions
+  },
+  {
+    "insert": "World",
+    "attribution": {                // the insertion "World" was deleted by Bob
+      "delete": [
+        "Bob"
+      ]
+    }
+  },
+  {
+    "insert": "attributions",       // the insertion "attributions" was inserted by Bob
```

Comment:
let's not use "attributions" in demo text because it's going to make the example code / json more confusing 😅

Reply:
Fair point ^^